src/shared/Core/Trace2Exception.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.ComponentModel;
+using System.IO;
+using GitCredentialManager.Authentication.OAuth;
+using GitCredentialManager.Interop;
+
+namespace GitCredentialManager;
+
+public class Trace2Exception : Exception
+{
+    public Trace2Exception(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+
+    public Trace2Exception(ITrace2 trace2, string message, string messageFormat) : base(message)
+    {
+        trace2.WriteError(message, messageFormat);
+    }
+}
+
+public class Trace2InvalidOperationException : InvalidOperationException
+{
+    public Trace2InvalidOperationException(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2OAuth2Exception : OAuth2Exception
+{
+    public Trace2OAuth2Exception(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2InteropException : InteropException
+{
+    public Trace2InteropException(ITrace2 trace2, string message, int errorCode) : base(message, errorCode)
+    {
+        trace2.WriteError($"message: {message} error code: {errorCode}");
+    }
+
+    public Trace2InteropException(ITrace2 trace2, string message, Win32Exception ex) : base(message, ex)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2GitException : GitException
+{
+    public Trace2GitException(ITrace2 trace2, string message, string gitMessage, int errorCode) :
+        base(message, gitMessage, errorCode)
+    {
+        trace2.WriteError($"message: {message} git message: {gitMessage} error code: {errorCode}");
```

Comment:
Do we know if the provided `gitMessage` is PII clean ??

I tracked it back and it looks like it is the contents of stderr of the child Git command.  I'm not sure what that contains in practice, so I'm just asking.

It looks like all of the callers pass a string literal for the `message`, so they should be safe.
I'm wondering if we should do something like:

```
string fmt = $"message: '{message}' error code: {errorCode} git message: '\{0\}'"
string msg = string.Format(fmt, gitMessage)

trace2.WriteError(msg, fmt)
```

Reply:
Probably a good idea.